alibi_detect/od/base.py

```diff
@@ -0,0 +1,32 @@
+from __future__ import annotations
+from typing import Union
+
+from typing_extensions import Protocol, runtime_checkable
+
+
+# Use Protocols instead of base classes for the backend associated objects. This is a bit more flexible and allows us to
+# avoid the torch/tensorflow imports in the base class.
+@runtime_checkable
+class TransformProtocol(Protocol):
+    """Protocol for transformer objects."""
+    def transform(self, x):
+        pass
+
+    def _transform(self, x):
+        pass
+
+
+@runtime_checkable
+class FittedTransformProtocol(TransformProtocol, Protocol):
+    """Protocol for fitted transformer objects."""
+    def fit(self, x_ref):
+        pass
+
+    def _fit(self, x_ref):
+        pass
+
+    def check_fitted(self):
+        pass
```

Comment:
Probably need a few docstrings for what each method is supposed to implement, especially the distinction between the private and public ones.

Any chance we can add things like argument and return types too or would be too cumbersome to work around the optional backend types?

Reply:
I've added links to the relevant backend objects. I feel this decreases the requirement for documentation and is more accurate since the above aren't the actual abstract classes themselves.

In terms of adding things like argument and return types my sense it that it could get messy. I'll create an issue to do so in a later PR.